clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -119,6 +123,11 @@ public Fetcher(ConsumerNetworkClient client,
      * @param cluster The current cluster metadata
      */
     public void initFetches(Cluster cluster) {
+        // Don't initiate fetch if we already have records available from previous fetchRequest
+        // This may happen when there is OffsetOutOfRange exception from previous fetchRequest
+        if (!records.isEmpty())
```

Comment:
Thanks for your patience as well.

Talking about your proposal, I am not convinced that we should avoid calling `poll()` in `pollOnce()` when data is available, for the following reasons.

1) Can you elaborate why the fact that fetchedRecords is not checked before blocking in poll() might already cause unnecessary blocking? In other word, under which sequence of events would it cause unnecessary blocking?

2) If we add the check for non-empty records in KafkaConsumer.pollOnce(), this would cause internal API change. We have agreed that this is a problem, where you proposed to use `Fetcher.outOfRangePartitions`.

What you described may actually be a valid problem and worth effort to fix. But I am not strong on this at this moment. Creation of a new ticket involving API change would require the creator to clearly describe the motivation and use case. I would rather leave it to a new ticket, to be taken by you or other developers, unless there is no other way to finish this ticket.

I would appreciate it to first get opinion on whether my latest proposal can solve the this ticket. Thank you much for review.